pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codestyle/AtLeastOneConstructorRule.java

```diff
@@ -0,0 +1,74 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.java.rule.codestyle;
+
+import java.util.Arrays;
+import java.util.Collection;
+
+import net.sourceforge.pmd.lang.java.ast.ASTClassOrInterfaceDeclaration;
+import net.sourceforge.pmd.lang.java.ast.ASTConstructorDeclaration;
+import net.sourceforge.pmd.lang.java.ast.ASTFieldDeclaration;
+import net.sourceforge.pmd.lang.java.ast.ASTMethodDeclaration;
+import net.sourceforge.pmd.lang.java.rule.AbstractIgnoredAnnotationRule;
+import net.sourceforge.pmd.lang.java.rule.design.UseUtilityClassRule;
+
+/**
+ * This rule detects non-static classes with no constructors;
+ * requiring even the default constructor to be explicit.
+ * It ignores classes with solely static methods,
+ * use {@link UseUtilityClassRule} to flag those.
+ */
+public class AtLeastOneConstructorRule extends AbstractIgnoredAnnotationRule {
+
+    public AtLeastOneConstructorRule() {
+        addRuleChainVisit(ASTClassOrInterfaceDeclaration.class);
+    }
+
+    @Override
+    protected Collection<String> defaultSuppressionAnnotations() {
+        return Arrays.asList("lombok.Data",
+                "lombok.Value",
+                "lombok.Builder",
+                "lombok.NoArgsConstructor",
+                "lombok.RequiredArgsConstructor",
+                "lombok.AllArgsConstructorAtLeastOneConstructor");
+    }
+
+    @Override
+    public Object visit(final ASTClassOrInterfaceDeclaration node, final Object data) {
+        // Ignore interfaces / static classes / classes that have a constructor / classes ignored through annotations
+        if (node.isInterface() || node.isStatic() || node.hasDescendantOfType(ASTConstructorDeclaration.class)
+                || hasIgnoredAnnotation(node)) {
+            return data;
+        }
+
+        boolean atLeastOneMember = false;
+
+        // Do we have any non-static methods?
+        for (final ASTMethodDeclaration m : node.findDescendantsOfType(ASTMethodDeclaration.class)) {
```

Comment:
Maybe you could condense these two loops into one by adding the method and field nodes to a single `List<AccessNode>`

Reply:
I considered that, but it would perform the 2 tree traversals every time even if not needed; so I desisted.

Reply:
Yes that makes sense. With Java 8 we could probably provide a lazy traversal Stream to do that kind of thing more concisely.